### DISCUSSION TITLE
Support for Changesets

### DISCUSSION THREAD

**Discussion by louiss0** [nrwl/nx#15566] [Feature Requests]
I wanted to talk about how nx did not have any good support for library workflows at all. Someone told me that there is a library that was created by the community. But I'm not fond of it. It forces me to use Git Hub conventions to use the system. I'm now asking for support for [Changesets](https://github.com/changesets/changesets). Whenever I say that a library is publishable I want to be prompted on whether or not I want to install Changesets. 

If you don't know what changesets is it's a library that was created to manage the publishing of libraries with the use of changelogs.

 **What it does is when you want to** 

- upgrade a library you use `npm changeset add`
- to puiblish a library  `pnpm changeset version`    

When you run changeset add you get prompted on which package you want to write a changeset for Then you'll get prompted for a message you write it down and after that, the message gets written down into a markdown file. When you want to publish the file then Change logs are created for you for each package.   

It even supports pre-releases and git tags.

**Comment by mxvsh** [comment]
Have you found any solution? I think a workaround is to just change the "outDir" of the packages to their package folder.

**Reply by louiss0** [reply]
No I think there is a thing called executors that could be used.

**Comment by ryanbas21** [comment]
@xencodes I think theres a few ways to do it but none are OOB Nx support. Changing the outdir is one way, and setting your `files` in package.json to support that outputted dist and ignoring the `dist/package.json` would probably work best.

Another way is to have a script which changes your workspaces to point to dist in your release process (but this has downsides with changelog being updated within the dist/ directory, you'd have to then move that to your folder and commit it back.

I would assume the first option is easiest. It would be best if there was support for an ourdir like `nx` has.

**Comment by Ambroos** [comment]
We recently got started with our Nx monorepo, and didn't want to invest too much time into switching release methods right away (and assumed changesets would just work with Nx). When it didn't I spent some time coming up with a workaround that might be helpful for others to make changesets work. It's definitely a hack, but it does work and doesn't require package patching. Here's the full setup, which involves package.json, project.json for any projects you want to support, a few scripts and customisations for the GH action. With this you just have to run the wrapped changeset script via npm/pnpm/yarn rather than running it directly / via npx. Args still work.

The strategy: creating a dummy lerna.json on the fly to make changesets detect package directories. There are two variants of this: one based on the source directories for the `version` command, and another one with dist directories (more hacky) for `publish`.

```json
{
  "name": "@org/your-monrepo",
  "version": "0.0.0",
  "private": true,
  "scripts": {
    "changeset": "node ./scripts/changeset-wrap.mjs",
    "publish-combo-changeset": "echo \"{}\" > lerna.json && pnpm exec nx run-many -t changeset-prepare-publish && pnpm exec changeset publish"
  }
```

```json
{
  "name": "my-lib",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/my-lib/src",
  "projectType": "library",
  "targets": {
    "build": {
      ...
    },
    "changeset-prepare-publish": {
      "cache": false,
      "executor": "nx:run-commands",
      "dependsOn": ["build"]
      "options": {
        "command": "node ./utils/changeset-publish-prepare.mjs dist/my-lib"
      }
    },
    ...
  }
}
```

```yml
# In your changeset version / release GH action
      - name: Hack Nx support into Changeset
        run: node ./scripts/changeset-ci-prepare.mjs

      - name: Changeset - create pull request / publish packages
        id: changesets
        uses: changesets/action@v1.4.5
        with:
          publish: pnpm publish-combo-changeset
          # the default version command is OK thanks to the action script executed before this
```

```
# .gitignore
lerna.json
```

```javascript
/**
 * create-src-lerna-json.mjs
 */
import nx from '@nx/devkit'

export default async function createSrcLernaJson() {
  const projectRoots = await nx
    .createProjectGraphAsync({ exitOnError: true, resetDaemonClient: true })
    .then((projectGraph) =>
      Object.entries(nx.readProjectsConfigurationFromProjectGraph(projectGraph).projects)
        .filter(([, p]) => {
          // Filter projects, we filter on a specific target existing. Easy to customise
          return 'changeset-prepare-publish' in (p.targets ?? {})
        })
        .map(([, p]) => `./${p.root}`),
    )

  const content = {
    ['not a real lerna.json']: 'This is a dummy lerna.json to support changeset',
    packages: projectRoots,
  }

  fs.writeFileSync('./lerna.json', JSON.stringify(content, null, 2))
  return content
}
```

```javascript
/**
 * changeset-ci-prepare.mjs
 */
import createSrcLernaJson from './create-src-lerna-json.mjs'
createSrcLernaJson().then((content) => {
  console.log('Created dummy lerna.json:', content)
})
```

```javascript
/**
 * changeset-wrap.mjs
 */
import fs from 'fs'
import { execSync } from 'child_process'
import createSrcLernaJson from './create-src-lerna-json.mjs'

// Verify we are in root by checking nx.json exists
if (!fs.existsSync('./nx.json')) {
  console.log('This script can only run from repo root.')
  process.exit(1)
}

createSrcLernaJson()
  .then(() => {
    // Run changeset, passing all args.
    const args = process.argv.slice(2).join(' ')
    execSync(`pnpm exec changeset ${args}`, { stdio: 'inherit' })
  })
  .then(() => {
    cleanup()
  })
  .catch((e) => {
    cleanup()
    // Exit with error
    console.error(e)
    process.exit(1)
  })

function cleanup() {
  // remove lerna.json
  fs.unlinkSync('./lerna.json')
}

process.on('SIGINT', function () {
  cleanup()
  process.exit()
})
```

```javascript
/**
 * add-to-lerna-json.mjs
 */

import fs from 'fs'

// Verify we are in root by checking nx.json exists
if (!fs.existsSync('./nx.json')) {
  console.log('This script can only run from repo root.')
  process.exit(1)
}

// get dist path from argv
const distPath = process.argv[2]

const lernaJson = JSON.parse(fs.readFileSync('./lerna.json', 'utf8'))
let existingProjects = lernaJson.packages ?? []

fs.writeFileSync(
  './lerna.json',
  JSON.stringify(
    {
      packages: [...existingProjects, distPath],
    },
    null,
    2,
  ),
)
```

**Comment by airtonix** [comment]
Have been using changesets for the last three years.

My two cents is that if your monorepo only contains libraries and not apps and this monorepo follows a fairly singular release path, then changesets might work for your team.

However here is the things I have found: 

- regarding developer UX
  - onboarding inexperienced developers creates friction i dont want.
  - most developers can't be bothered writing good changeset notes
  - they have to fill out a PR description anyway

- regarding release tooling:
  - doesn't support apps with different release schedules

I've spent a great deal of time looking into alternatives for a monorepo that has apps and libs that have concurrent release schedules.

So far my favourite alternative seems to be googles release-please if i can get it to keep and update concurrent release prs.
